api/v1/history_types.go

```diff
@@ -0,0 +1,137 @@
+	}
+	return &h[0]
+}
+
+// Truncate keeps only the latest snapshots in the history up to MaxHistorySize.
+// Since the history is maintained with most recent first, we simply truncate from the end.
+func (h *History) Truncate() {
+	if len(*h) <= MaxHistorySize {
+		return
+	}
+	*h = (*h)[:MaxHistorySize]
+}
+
+// Upsert adds a new snapshot to the history or updates an existing one
+// with the same digest and status. The most recent snapshot is moved to the front of the array.
+// When adding new snapshots, the history is automatically truncated to MaxHistorySize.
+func (h *History) Upsert(digest string, timestamp time.Time, duration time.Duration, status string, metadata map[string]string) {
+	now := metav1.NewTime(timestamp)
+	durationMeta := metav1.Duration{Duration: duration}
+
+	// Look for existing snapshot with same digest
+	for i := range *h {
+		if (*h)[i].Digest == digest && (*h)[i].LastReconciledStatus == status {
+			// Update existing snapshot
+			(*h)[i].LastReconciled = now
+			(*h)[i].LastReconciledDuration = durationMeta
+			(*h)[i].TotalReconciliations++
+			(*h)[i].Metadata = metadata
+			// Move to front if not already there
+			if i > 0 {
+				snapshot := (*h)[i]
+				copy((*h)[1:i+1], (*h)[0:i])
+				(*h)[0] = snapshot
+			}
+			return
+		}
+	}
+
+	// Add new snapshot at the front
+	newSnapshot := Snapshot{
+		Digest:                 digest,
+		FirstReconciled:        now,
+		LastReconciled:         now,
+		LastReconciledDuration: durationMeta,
+		LastReconciledStatus:   status,
+		TotalReconciliations:   1,
+		Metadata:               metadata,
+	}
+
+	*h = append([]Snapshot{newSnapshot}, *h...)
+	h.Truncate()
+}
+
+// Snapshot represents a point-in-time record of a group of resources reconciliation,
+// including timing information, status, and a unique digest identifier.
+type Snapshot struct {
+	// Digest is the checksum in the format `<algo>:<hex>` of the resources in this snapshot.
+	// +required
+	Digest string `json:"digest"`
+
+	// FirstReconciled is the time when this revision was first reconciled to the cluster.
+	// +required
+	FirstReconciled metav1.Time `json:"firstReconciled"`
+
+	// LastReconciled is the time when this revision was last reconciled to the cluster.
+	// +required
+	LastReconciled metav1.Time `json:"lastReconciled"`
+
+	// LastReconciledDuration is time it took to reconcile the resources in this revision.
+	// +kubebuilder:validation:Type=string
+	// +required
+	LastReconciledDuration metav1.Duration `json:"lastReconciledDuration"`
+
+	// LastReconciledStatus is the status of the last reconciliation.
+	// +required
+	LastReconciledStatus string `json:"lastReconciledStatus"`
+
+	// TotalReconciliations is the total number of reconciliations that have occurred for this snapshot.
+	// + required
+	TotalReconciliations int64 `json:"totalReconciliations"`
+
+	// Metadata contains additional information about the snapshot.
+	// +optional
+	Metadata map[string]string `json:"metadata,omitempty"`
```

Comment:
Is there a good reason for using `map[string]string` here? Do we expect the content of this field to change over time? From what I can see, the same fields are passed everywhere and both are numbers, we could just have a `SnapshotMetadata` struct for that. I find it really bad design to strings for representing numbers, unless we have a strong reason for this that I miss the context for.

Reply:
The History is not specific to RSET, the FluxInstance will store here the Flux version, RSIP will store commit sha/digest, etc. This type matches the Kubernetes meta annotations, it's for storing controller-specific info without leaking all these details in the History struct which is meant to be generic.